sherpa/astro/io/xstable.py

```diff
@@ -0,0 +1,752 @@
+each bin - have units of photon/cm^2/s. This is easy to accidentally
+change - e.g.  in the example above if mdl.norm were changed to a
+value other than 1 everything would work but Sherpa and XSPEC would
+infer incorrect fluxes or luminosities.
+
+"""
+
+from dataclasses import dataclass, field
+import itertools
+import time
+from typing import Any, Optional, Union
+
+import numpy as np
+
+import sherpa
+
+
+__all__ = ("IntParam", "Param", "make_xstable_model",
+           "write_xstable_model")
+
+
+# In the typing support I have used Any to mean "ndarray".
+#
+
+@dataclass
+class Param:
+    """Represent a parameter.
+
+    The absolute DELTA value is used by XSPEC but ignored by Sherpa.
+    The sign of DELTA is used (by both Sherpa and XSPEC) to decide if
+    the parameter should be frozen (DELTA is negative) or not by
+    default.
+
+    """
+    name: str
+    """The parameter name.
+
+    This must be unique and ideally a valid attribute name in Python,
+    although the latter is not enforced.
+    """
+    initial: float
+    """The default value for the parameter."""
+    delta: float
+    """The delta value for fitting.
+
+    If negative then the parameter defaults to being frozen. Sherpa
+    does not use this value (other than checking the sign) but XSPEC
+    does.
+    """
+    hardmin: float
+    """The minimum value for the parameter."""
+    hardmax: float
+    """The maximum value for the parameter."""
+    softmin: Optional[float] = None
+    """The "soft" minimum. Defaults to hardmin if not given.."""
+    softmax: Optional[float] = None
+    """The "soft" maximum. Defaults to hardmax if not given."""
+
+    def __post_init__(self):
+        """Validate the settings"""
+
+        # Do we need to set up softmin/max?
+        #
+        if self.softmax is None:
+            self.softmax = self.hardmax
+
+        if self.softmin is None:
+            self.softmin = self.hardmin
+
+        # There should probably be a check that the name is not
+        # an invalid Python attribute name.
+        #
+        if self.name.strip() == "":
+            raise ValueError("name can not be empty")
+
+        if self.softmin < self.hardmin:
+            raise ValueError(f"Parameter {self.name} softmin={self.softmin} < hardmin={self.hardmin}")
+
+        if self.softmax > self.hardmax:
+            raise ValueError(f"Parameter {self.name} softmax={self.softmax} > hardmax={self.hardmax}")
+
+        # We allow min == max
+        if self.softmin > self.softmax:
+            raise ValueError(f"Parameter {self.name} softmin={self.softmin} > softmax={self.softmax}")
+
+        if self.initial < self.softmin or self.initial > self.softmax:
+            raise ValueError(f"Parameter {self.name} initial={self.initial} outside softmin={self.softmin} to softmax={self.softmax}")
+
+        if self.delta == 0:
+            raise ValueError("delta can not be 0")
+
+
+@dataclass
+class IntParam(Param):
```

Comment:
I read this as "IntegerParam". Can we call it "InterpolatedParam"?

Reply:
As I expect this class to be used more than the Param class I'm renamnig things to

    Param -> BsaeParam
    IntParam -> Param

since `InterpolatedParameter` is a better name but it's too long...

Reply:
I've gone with my renaming.